src/relay/pass/partial_eval.cc

```diff
@@ -0,0 +1,714 @@
+  }
+  template <typename T>
+  T* try_get() {
+    auto ret = dynamic_cast<T*>(this);
+    return ret;
+  }
+};
+
+using Static = std::shared_ptr<StaticNode>;
+
+struct PStaticNode;
+
+using PStatic = std::shared_ptr<PStaticNode>;
+
+struct STupleNode : StaticNode {
+  std::vector<PStatic> fields;
+  explicit STupleNode(const std::vector<PStatic>& fields) : fields(fields) { }
+};
+
+Static STuple(const std::vector<PStatic>& fields) {
+  return std::make_shared<STupleNode>(fields);
+}
+
+struct STensorNode : StaticNode {
+  runtime::NDArray data;
+  explicit STensorNode(const NDArray& data) : data(data) { }
+};
+
+Static STensor(const NDArray& data) {
+  return std::make_shared<STensorNode>(data);
+}
+
+struct SConstructorNode : StaticNode {
+  Constructor constructor;
+  std::vector<PStatic> fields;
+  SConstructorNode(const Constructor& constructor, const std::vector<PStatic>& fields) :
+    constructor(constructor), fields(fields) { }
+};
+
+Static SConstructor(const Constructor& constructor, const std::vector<PStatic>& fields) {
+  return std::make_shared<SConstructorNode>(constructor, fields);
+}
+
+struct SRefNode : StaticNode { };  // we will use the pointer as the guid for hashing
+
+Static SRef() {
+  return std::make_shared<SRefNode>();
+}
+
+using Clos = std::function<PStatic(const std::vector<PStatic>&,
+                                   const Attrs&,
+                                   const Array<Type>&,
+                                   LetList*)>;
+
+struct SClosNode : StaticNode {
+  Clos func;
+  explicit SClosNode(const Clos& func) : func(func) { }
+};
+
+Static SClos(const Clos& func) {
+  return std::make_shared<SClosNode>(func);
+}
+
+struct PStaticNode {
+  Static pstatic;  // may be null
+  Expr dynamic;
+  PStaticNode(const Static& pstatic, const Expr& dynamic) : pstatic(pstatic), dynamic(dynamic) { }
+  explicit PStaticNode(const Expr& dynamic) : PStaticNode(Static(), dynamic) { }
+};
+
+/*!
+ * \brief A stack frame in the Relay interpreter.
+ *
+ * Contains a mapping from relay::Var to relay::Value.
+ */
+struct Frame {
+  /*! \brief The set of local variables and arguments for the frame. */
+  std::unordered_map<Var, PStatic, NodeHash, NodeEqual> locals;
+  explicit Frame(const std::unordered_map<Var, PStatic, NodeHash, NodeEqual>& locals) :
+    locals(locals) { }
+  Frame() = default;
+};
+
+class Environment {
+ public:
+  Environment() : env_({Frame()}) { }
+
+  template<typename T>
+  T Extend(const std::function<T()>& cont) {
+    RAII raii(this);
+    return cont();
+  }
+
+  void Insert(const Var& v, const PStatic& ps) {
+    env_.back().locals[v] = ps;
+  }
+
+  // return null if not found
+  PStatic Lookup(const Var& v) {
+    auto rit = env_.rbegin();
+    while (rit != env_.rend()) {
+      if (rit->locals.find(v) != rit->locals.end()) {
+        return rit->locals.find(v)->second;
+      }
+      ++rit;
+    }
+    return PStatic();
+  }
+
+ private:
+  std::list<Frame> env_;
+
+  struct RAII {
```

Comment:
I think that this might be more aptly named `FrameContext` than the generic term `RAII`